<p><img src="https://files.cdn.thinkific.com/file_uploads/133292/images/a9e/c1a/bc3/P02.jpg"></p>

`MCSDSC02P01V3______`

`MCSDSC02P01V2______`

`MCSDSELIMC02P01V1010621`

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Отлично! Поздравляем, проект принят!

Местами оставил комментарии под заголовком `Комментарий ассистента 2`

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Хорошая работа! Критических замечаний почти нет, но есть доработки.

Как обычно, заголовки (комментариев) будут трех цветов. <font color="89DD00">Зеленый</font> — все правильно (и если комментария нет, то это также значит, что все правильно). <font color="FFB800">Желтый</font> — можно доработать. <font color="FF0F00">Красный</font> — нужно исправить.

Постарайся учесть эти комментарии в ходе доработки проекта и не удаляй их.

## Постановка задачи

Без очереди — это мобильное приложение, которое помогает сделать заказ до прихода в кафе, чтобы не ждать официанта, приготовление еды и счет.

Необходимо изучить данные сервиса с тестового запуска на территории дизайн-завода Флакон и Хлебозавода №9 и ответить на вопросы, поставленные менеджером проекта:

    1. На какие кафе и на какие дни приходится больше всего заказов?

    2. В какое время приложением пользуются чаще всего?

    3. Насколько большими компаниями приходят в кафе?


<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Постановка задачи в порядке

## План работы

1) Предобработка данных (переименовать столбцы, удалить дубликаты, заполнить пропуски,
 исправить ошибки в названиях ресторанов)

2) Проанализировать распределение заказов в течение недели (вычислить, на какие кафе приходится больше всего заказов,
написать функцию для вычисления количества заказов в определенный день, свести результаты
в таблицу, в которой столбцами являются ресторан и 3 дня недели)

3) Проанализировать распределение заказов в течение дня (написать функцию, которая возвращает количество заказов
 за определенный промежуток времени (обеденное и вечернее))

4) Вычислить типичное количество человек за столиком
 (написать функцию, которая определяет моду (типичное значение) количества человек, для которых заказывается столик.)

5) Общие выводы

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Отлично

## Импорт данных и библиотек

In [41]:
import pandas as pd
import numpy as np
import datetime

Прочитаем файл `bez_ocheredi.csv` и сохраним его в переменную `bo_data`.

In [42]:
bo_data = pd.read_csv('../../data/bez_ocheredi.csv')

Получим первые 10 строк таблицы.

In [43]:
bo_data.head(10)

userID    ресторан     weekday  ordered    Столик для 
0  b912c3c6a6ad   Iskra Bar     Monday     10:00             1
1  2dac0150717d   Iskra Bar     Monday     13:30             2
2  ba83aefab5dc   Iskra Bar    Tuesday     12:30             3
3  02996f1dd2ea   Iskra Bar  Wednesday     12:00             2
4  7c90be56d2ab   Iskra Bar   Thursday     14:30             3
5  51e0962387f7      Там Ем     Friday     12:30             2
6  c4f260a2b48a         NaN   Saturday     15:00             6
7  6b71309d6a8a  Крафтейнер     Monday     10:30             1
8  c5b969cb63a2  Бар Сибирь  Wednesday     15:30             2
9  b1570962e68c  Бар Сибирь  Wednesday     12:30             3

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Комментарий учтен

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Для вывода первых строк в pandas лучше применять head(), передав нужное количество строк в аргумент

Получим общую информацию о данных из таблицы.

In [44]:
bo_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1024 entries, 0 to 1023
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   userID        1024 non-null   object
 1    ресторан     959 non-null    object
 2    weekday      1024 non-null   object
 3    ordered      1024 non-null   object
 4    Столик для   1024 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 40.1+ KB


Столбцы в `bo_data` содержат следующую информацию:

* userID — уникальный ID пользователя
* ресторан — название ресторана
* weekday — день недели
* ordered — на какое время сделан заказ
* Столик для — для скольких людей заказан столик

### Выводы

* Данные содержат строковые значения и числа
* Не все данные записаны в подходящем формате, например время следовало бы записать в формате datetime
* Данные содержат пропуски. Это затрудняет дальнейший анализ

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Комментарий учтен

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

А еще в наших данных есть пропущенные значения в колонке "ресторан", об этом тоже необходимо указать в выводах, поскольку это существенно влияет на дальнейший анализ

По поводу datetime - в этом проекте мы еще не закладывали применение этого формата, здесь можно все сделать и без него. Но если он применен, то это только приветствуется)

## Предобработка данных

Получим названия столбцов.

In [45]:
bo_data.columns.values.tolist()

['userID ', ' ресторан ', ' weekday', ' ordered ', ' Столик для ']

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Можно получить наименования столбцов гораздо проще - `bo_data.columns`

Обращаем внимание, что помимо названий на разном языке, в них есть пробелы, которые затрудняют доступ к данным. Переименуем столбцы и проверим результат.

In [46]:
labels = ['user_id', 'restaurant', 'weekday', 'order_time', 'number_of_people']
bo_data.set_axis(labels, axis='columns', inplace=True)

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Комментарий учтен, теперь колонки переименованы корректно

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Не надо переименовывать столбцы через атрибут columns, потому что так ты неявно учитываешь порядок столбцов. Если порядок поменяется, то код не будет работать. 

Все столбцы можно переименовать с помощью метода `set_axis()`. Передавай ему список с названиями столбцов, `axis='columns'`, `inplace = True` (последний аргумент — меняет структуру данных, а не создает копию). Также можно использовать метод `rename()`, но опять же указывай `inplace = True`.

Изменение формата данных столбца order_time на datetime

In [47]:
bo_data['order_time'] = pd.to_datetime(bo_data['order_time'], format= '%H:%M' ).dt.time

In [48]:
bo_data.head()

user_id restaurant    weekday order_time  number_of_people
0  b912c3c6a6ad  Iskra Bar     Monday   10:00:00                 1
1  2dac0150717d  Iskra Bar     Monday   13:30:00                 2
2  ba83aefab5dc  Iskra Bar    Tuesday   12:30:00                 3
3  02996f1dd2ea  Iskra Bar  Wednesday   12:00:00                 2
4  7c90be56d2ab  Iskra Bar   Thursday   14:30:00                 3

Еще раз проверим данные на наличие пропусков, просуммировав их.

In [49]:
bo_data.isna().sum()

user_id              0
restaurant          65
weekday              0
order_time           0
number_of_people     0
dtype: int64

Менеджер проекта предупреждал, что некоторые заказы заводились на тестовую карточку кафе без названия, поэтому есть смысл заполнить такой пропуск словом admin.

In [50]:
bo_data['restaurant'].fillna('admin', inplace=True)

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Отлично, комментарий учтен

<h3><font color="#FF0F00">Комментарий ассистента</font></h3>

Без указания конкретного столбца на `admin` заменятся все пропуски всех колонок. В нашем случае пропуски только в одном столбце и разницы мы не увидим, но необходимо предусмотреть случай пропусков еще и в других колонках.
Укажи столбец, в котором заполняются пропуски

Проверяем, что пропусков не осталось.

In [51]:
bo_data.isna().sum()

user_id             0
restaurant          0
weekday             0
order_time          0
number_of_people    0
dtype: int64

Установим количество явных дубликатов. Если найдутся, то удалим и снова проверим, что их не осталось.

In [52]:
bo_data.duplicated().sum()

24

In [53]:
bo_data = bo_data.drop_duplicates().reset_index(drop = True)

In [54]:
bo_data.duplicated().sum()

0

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

По дубликатам всё отлично

Скорее всего, такие полные дубликаты говорят о сбое при выгрузке данных. Необходимо сообщить об этом.

Еще одна проблема, о которой предупреждал менеджер проекта — дубликаты карточки Iskra Bar с неправильными названием бара - "Искра" и "Искр". На них заводились реальные заказы, поэтому необходимо вычислить их и переименовать на правильное название — Iskra Bar.

Объявим функцию `replace_restaurant()`, которая заменяет неправильное название, а потом проверяет успешность выполнения замены подсчетом (если замена прошла успешно — результат подсчета должен быть 0).

In [55]:
def replace_restaurant(df, wrong_name, correct_name):
    return df.replace(wrong_name, correct_name)


Передаем функции названия 'Искра' и 'Искр'.

In [56]:
bo_data = replace_restaurant(bo_data, 'Искр', 'Iskra Bar')

In [57]:
bo_data = replace_restaurant(bo_data, 'Искра', 'Iskra Bar')

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Отлично! Функция работает корректно

### Выводы

Данные были не очень чистыми, следует доработать систему их сбора со стороны компании

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Комментарий учтен

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

"не были не очень чистыми" - из-за двойного отрицания путаница)  
Человек, который пропустит код и сразу посмотрит выводы сначала будет долго думать что имелось ввиду

<font color="#800080">очепятка)</font>

## Распределение заказов в течение недели

Считаем, на какие кафе приходится больше всего заказов.

In [58]:
bo_data['restaurant'].value_counts()

Iskra Bar     287
Там Ем        200
Бар Сибирь    159
Северия       107
Крафтейнер     78
Коворкафе      66
Азия           62
admin          41
Name: restaurant, dtype: int64

Группируем по `weekday` и считаем, на какие дни приходится больше всего заказов.

In [59]:
bo_data['weekday'].value_counts()

Tuesday      251
Wednesday    242
Thursday     180
Monday       141
Friday       110
Saturday      52
Sunday        24
Name: weekday, dtype: int64

Объявляем функцию, которая возвращает количество заказов для ресторана в определенный день.

In [60]:
def orders_per_weekday(restaurant, weekday):
    return bo_data[(bo_data['restaurant'] == restaurant) & (bo_data['weekday'] == weekday)]['restaurant'].count()

Передадим ей заведения Iskra Bar и Там Ем — два флагманских места, по которым просили результаты прежде всего — и понедельник, среду и пятницу в качестве дня недели.

In [61]:
mon_iskra = orders_per_weekday('Iskra Bar', 'Monday')
mon_iskra

48

In [62]:
wed_iskra = orders_per_weekday('Iskra Bar', 'Wednesday')
wed_iskra

53

In [63]:
fri_iskra = orders_per_weekday('Iskra Bar', 'Friday')
fri_iskra

29

In [64]:
mon_tamem = orders_per_weekday('Там Ем', 'Monday')
mon_tamem

25

In [65]:
wed_tamem = orders_per_weekday('Там Ем', 'Wednesday')
wed_tamem

52

In [66]:
fri_tamem = orders_per_weekday('Там Ем', 'Friday')
fri_tamem

24

Сведем результат в таблицу.

In [67]:
restaurants = ['Iskra Bar', 'Там Ем']
weekdays = ['Monday', 'Wednesday', 'Friday']
orders = np.array([mon_iskra, wed_iskra, fri_iskra, mon_tamem, wed_tamem, fri_tamem])

In [68]:
result_weekdays = pd.DataFrame(index=restaurants, columns=weekdays, data=orders.reshape(2, 3))
result_weekdays

Monday  Wednesday  Friday
Iskra Bar      48         53      29
Там Ем         25         52      24

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Всё супер! `numpy` здесь мы еще не планировали использовать, но респект за уместное и правильное применение :)

### Выводы

* Самые популярные рестораны: "Iskra Bar" и "Там Ем"
* Больше всего заказов приходится на середину недели: среду и четверг
* Меньше всего заказов в выходные. Связано это может быть с тем, что в выходные люди меньше торопятся, и предпочитают закаазывать еду на месте

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Всё так. Действительно, в выходные люди могут больше заказывать на месте, но еще можно допустить, что эти места популярны и офисных работников, которые любят обедать, используя приложение

## Распределение заказов в течение дня

Сравним количества заказов, сделанных в обеденное и вечернее время — с 12 до 16 и с 16 до 22. Для этого объявим функцию, которая принимает данные и 2 аргумента времени.

In [69]:
def orders_in_time(df, hour_from, hour_to):
    return df['restaurant'][(df['order_time'] >= datetime.time(hour_from, 0, 0)) & (df['order_time'] < datetime.time(hour_to, 0, 0))].count()

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Комментарий учтен, функция корректная

<h3><font color="#FFB800">Комментарий ассистента</font></h3>
 
Результаты функции правильные, но применение `shape[0]` здесь некорректно.  
Лучше использовать `count()` и указать, например, колонку по которой выводится количество:

`return df['restaurant'][(df['order_time'] >= datetime.time(hour_from, 0, 0)) & (df['order_time'] < datetime.time(hour_to, 0, 0))].count()`

И как я писал выше, здесь можно было не использовать datetime, и тогда функцию можно было бы сделать намного легче:

`def orders_in_time(df, hour_from, hour_to):
    return df['restaurant'][(df['ordered_time'] >= hour_from) & (df['ordered_time'] < hour_to)].count()   `

Передадим промежуток с 12:00 и 16:00

In [70]:
orders_in_time(bo_data, 12, 16)

543

И с 16:00 до 22:00

In [71]:
orders_in_time(bo_data, 16, 22)

338

### Выводы

* Заказов в дневное время больше, чем в вечернее. Связано это может быть с тем, что в течение рабочего дня люди больше торопятся (чтобы успеть на встречу, либо чтобы уложиться в обеденный перерыв)

<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Всё верно, в обеденное время столиков бронируется больше

## Типичное количество человек за столиком

Выясним, для какого количества человек обычно заказывают столик в разное время на примере Iskra Bar и Там Ем. Для этого напишем функцию, которая принимает данные, название ресторана, 2 времени и считает моду.

In [72]:
def people_in_time(df, restaurant, hour_from, hour_to):
    return df[(df['restaurant'] == restaurant) &
              (df['order_time'] >= datetime.time(hour_from, 0, 0)) &
              (df['order_time'] < datetime.time(hour_to, 0, 0))]['number_of_people'].mode()


<h3><font color="#89DD00">Комментарий ассистента</font></h3>

Отлично, функция корректная

Проверим функцию на двух ресторанах, а также дневном и вечернем промежутке времени.

In [73]:
people_in_time(bo_data, 'Iskra Bar', 12, 16)

0    2
dtype: int64

In [74]:
people_in_time(bo_data, 'Iskra Bar', 16, 22)

0    1
dtype: int64

In [75]:
people_in_time(bo_data, 'Там Ем', 12, 16)

0    2
dtype: int64

In [76]:
people_in_time(bo_data, 'Там Ем', 16, 22)

0    1
dtype: int64

### Выводы

* Днем люди обычно приходят вдвоем, вечером же по одиночке. Это может быть связано с тем, что днем люди могут чаще обедать с коллегами по работе

## Общие выводы

* Приложению необходимо доработать сбор данных
* Приложением чаще пользуются в будние днем. Связано это может быть с тем, что
людям необходимо экономить время в рабочие время. В выходные же посетители не против
подождать чек и посидеть лишние 10 минут за столиком
* Существует достаточно большой разброс между количеством заказов в популярных ресторанах
и в более мелких (Iskra Bar: 287, Азия: 62)
* Обычно заказ делают на одного-двух человек

<h3><font color="#89DD00">Комментарий ассистента 2</font></h3>

Комментарии учтены, отличная работа!

<h3><font color="#FFB800">Комментарий ассистента</font></h3>

Все результаты и твои выводы правильные, отлично!

Осталось внести небольшие правки (см. комментарии), и ждем проект на финальное ревью :)